vllm/model_executor/layers/quantization/awq_triton.py

```diff
@@ -0,0 +1,549 @@
+    zeros = reverse_awq_order(zeros)
+
+    iweights = reverse_awq_order(iweights)
+
+    iweights = torch.bitwise_and(iweights, (2**bits) - 1)
+    zeros = torch.bitwise_and(zeros, (2**bits) - 1)
+
+    scales = scales.repeat_interleave(group_size, dim=0)
+    zeros = zeros.repeat_interleave(group_size, dim=0)
+    return (iweights - zeros) * scales, zeros
+
+
+def test_dequantize():
+    print("=" * 10 + " TESTING DEQUANTIZE" + "=" * 10)
+    use_triton = True
+    use_torch = True
+
+    qweight_rows = 3584
+    qweight_cols = 576
+    group_size = 128
+    small_test_size = False
+    if small_test_size:
+        qweight_rows = 256
+        qweight_cols = 128
+    print(f"qweight_rows = {qweight_rows}, qweight_cols = {qweight_cols}")
+    qweight_dtype = torch.int32
+    scales_rows = qweight_rows // group_size
+    scales_cols = qweight_cols * 8
+    scales_dtype = torch.float16
+    zeros_rows = scales_rows
+    zeros_cols = qweight_cols
+    zeros_dtype = torch.int32
+    split_k_iters = 0
+    thx = 0
+    thy = 0
+    torch.manual_seed(0)
+
+    qweight = torch.randint(0,
+                            10000000, (qweight_rows, qweight_cols),
+                            dtype=qweight_dtype,
+                            device=device)
+    scales = torch.rand(scales_rows,
+                        scales_cols,
+                        dtype=scales_dtype,
+                        device=device)
+    zeros = torch.randint(0,
+                          10000000, (zeros_rows, zeros_cols),
+                          dtype=zeros_dtype,
+                          device=device)
+    print(f"qweight = {qweight}")
+    if use_triton:
+        iweights_triton = awq_dequantize_triton(qweight, scales, zeros,
+                                                split_k_iters, thx, thy)
+        print(f"Triton result:iweights_triton = {iweights_triton}")
+        print("Any infs in triton result? -->"
+              f"{torch.any(torch.isinf(iweights_triton))}")
+
+    if use_torch:
+        iweights_torch, _ = awq_dequantize_torch(qweight, scales, zeros,
+                                                 split_k_iters, thx, thy)
+        print(f"Torch result:iweights_torch = {iweights_torch}")
+
+    if use_torch and use_triton:
+        diff = iweights_torch - iweights_triton
+        error = torch.sum(torch.sqrt(diff * diff))
+        print(f"error = {error}")
+
+
+def awq_gemm_triton(input: torch.Tensor, qweight: torch.Tensor,
+                    scales: torch.Tensor, qzeros: torch.Tensor,
+                    split_k_iters: int) -> torch.Tensor:
+    M, K = input.shape
+    N = qweight.shape[1] * 8
+    grid = lambda META: (
+        triton.cdiv(M, META['BLOCK_SIZE_M']) * triton.cdiv(
+            N, META['BLOCK_SIZE_N']),
+        split_k_iters,
+    )
+    awq_group_size = 128
+    block_size_m = 32
+    block_size_n = 32
+    block_size_k = 32
+
+    result = torch.zeros((M, N), dtype=torch.float16, device=input.device)
```

Comment:
Could you add support for bf16 as well, if not too much effort?

Reply:
I could try, I will talk to James about this.  I'm not sure if AWQ itself supports bf16, but I will look into it.

Reply:
Should support bf16 now